File: app/models/custom_field.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Index, Integer, String, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class CustomFieldDefinition(Base):
    __tablename__ = "custom_field_definitions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    client_id: Mapped[int] = mapped_column(
        ForeignKey("clients.id", ondelete="CASCADE"), nullable=False
    )
    entity_type: Mapped[str] = mapped_column(String(20), nullable=False)  # "lead" | "company"
    field_key: Mapped[str] = mapped_column(String(100), nullable=False)
    field_label: Mapped[str] = mapped_column(String(255), nullable=False)
    field_type: Mapped[str] = mapped_column(String(50), nullable=False)  # text|number|date|boolean|select
    options: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    is_required: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    show_in_table: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    # Enrichment auto-fill: source provider name (e.g. "apollo") and dot-notation path within
    # that provider's response (e.g. "employee_count").  Both must be set together.
    enrichment_source: Mapped[str | None] = mapped_column(String(50), nullable=True)
    enrichment_mapping: Mapped[str | None] = mapped_column(String(200), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    __table_args__ = (
        UniqueConstraint(
            "client_id", "entity_type", "field_key",
            name="uq_custom_field_client_entity_key",
        ),
        Index("ix_custom_field_client_id", "client_id"),
        Index("ix_custom_field_entity_type", "entity_type"),
    )

File: frontend/src/types/custom_field.ts
export interface CustomFieldDefinition {
  id: string
  client_id: number
  entity_type: "lead" | "company"
  field_key: string
  field_label: string
  field_type: "text" | "number" | "date" | "boolean" | "select"
  options: string[] | null
  is_required: boolean
  show_in_table: boolean
  sort_order: number
  enrichment_source: string | null
  enrichment_mapping: string | null
  created_at: string
  updated_at: string
}

export type CustomFieldValues = Record<string, string | number | boolean | null>

export interface CustomFieldDefinitionCreate {
  entity_type: "lead" | "company"
  field_key: string
  field_label: string
  field_type: "text" | "number" | "date" | "boolean" | "select"
  options?: string[] | null
  is_required?: boolean
  show_in_table?: boolean
  sort_order?: number
  enrichment_source?: string | null
  enrichment_mapping?: string | null
}

export interface CustomFieldDefinitionUpdate {
  field_label?: string
  field_type?: "text" | "number" | "date" | "boolean" | "select"
  options?: string[] | null
  is_required?: boolean
  show_in_table?: boolean
  sort_order?: number
  enrichment_source?: string | null
  enrichment_mapping?: string | null
}
